File: app/src/main/java/xyz/zedler/patrick/tack/util/NotificationUtil.java
/*
 * This file is part of Tack Android.
 *
 * Tack Android is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Tack Android is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Tack Android. If not, see http://www.gnu.org/licenses/.
 *
 * Copyright (c) 2020-2025 by Patrick Zedler
 */

package xyz.zedler.patrick.tack.util;

import android.Manifest;
import android.annotation.SuppressLint;
import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.os.Build.VERSION;
import android.os.Build.VERSION_CODES;
import androidx.annotation.NonNull;
import androidx.appcompat.view.ContextThemeWrapper;
import androidx.core.app.NotificationCompat;
import androidx.core.app.NotificationCompat.Action;
import androidx.core.app.NotificationManagerCompat;
import androidx.core.content.ContextCompat;
import com.google.android.material.color.DynamicColors;
import xyz.zedler.patrick.tack.Constants.ACTION;
import xyz.zedler.patrick.tack.Constants.DEF;
import xyz.zedler.patrick.tack.Constants.PREF;
import xyz.zedler.patrick.tack.Constants.THEME;
import xyz.zedler.patrick.tack.R;
import xyz.zedler.patrick.tack.activity.MainActivity;
import xyz.zedler.patrick.tack.service.MetronomeService;

public class NotificationUtil {

  private static final String TAG = NotificationUtil.class.getSimpleName();

  private final static String CHANNEL_ID = "metronome";
  private final static int REQUEST_CODE = 0;
  public final static int NOTIFICATION_ID = 1;

  private final Context context;
  private final SharedPreferences sharedPrefs;
  private final NotificationManagerCompat notificationManager;

  public NotificationUtil(Context context) {
    this.context = context;
    sharedPrefs = new PrefsUtil(context).getSharedPrefs();
    notificationManager = NotificationManagerCompat.from(context);
  }

  public boolean hasPermission() {
    return hasPermission(context);
  }

  public static boolean hasPermission(@NonNull Context context) {
    if (VERSION.SDK_INT >= VERSION_CODES.TIRAMISU) {
      int status = ContextCompat.checkSelfPermission(
          context, Manifest.permission.POST_NOTIFICATIONS
      );
      return status == PackageManager.PERMISSION_GRANTED;
    } else {
      return true;
    }
  }

  public void createNotificationChannel() {
    if (VERSION.SDK_INT < VERSION_CODES.O) {
      return;
    }
    notificationManager.createNotificationChannel(
        new NotificationChannel(
            CHANNEL_ID,
            context.getString(R.string.title_notification_channel),
            NotificationManager.IMPORTANCE_HIGH
        )
    );
  }

  @SuppressLint("MissingPermission")
  public void updateNotification(Notification notification) {
    if (hasPermission()) {
      notificationManager.notify(NOTIFICATION_ID, notification);
    }
  }

  public Notification getNotification(boolean playButton) {
    Intent openIntent = new Intent(context, MainActivity.class);
    openIntent.setFlags(Intent.FLAG_ACTIVITY_SINGLE_TOP | Intent.FLAG_ACTIVITY_REORDER_TO_FRONT);
    PendingIntent activityPendingIntent = PendingIntent.getActivity(
        context, REQUEST_CODE, openIntent, PendingIntent.FLAG_IMMUTABLE
    );

    Intent startIntent = new Intent(context, MetronomeService.class);
    startIntent.setAction(ACTION.START);
    PendingIntent startServicePendingIntent = PendingIntent.getService(
        context, REQUEST_CODE, startIntent, PendingIntent.FLAG_IMMUTABLE
    );
    Action actionStart = new Action(
        R.drawable.ic_rounded_play_arrow_fill,
        context.getString(R.string.action_play),
        startServicePendingIntent
    );

    Intent stopIntent = new Intent(context, MetronomeService.class);
    stopIntent.setAction(ACTION.STOP);
    PendingIntent stopServicePendingIntent = PendingIntent.getService(
        context, REQUEST_CODE, stopIntent, PendingIntent.FLAG_IMMUTABLE
    );
    Action actionStop = new Action(
        R.drawable.ic_rounded_stop_fill,
        context.getString(R.string.action_stop),
        stopServicePendingIntent
    );

    String title = context.getString(R.string.msg_service_running);
    String text = context.getString(R.string.msg_service_running_return);
    return new NotificationCompat.Builder(context, CHANNEL_ID)
        .setContentTitle(title)
        .setContentText(text)
        .setStyle(new NotificationCompat.BigTextStyle().setBigContentTitle(title).bigText(text))
        .setContentIntent(activityPendingIntent)
        .addAction(playButton ? actionStart : actionStop)
        .setAutoCancel(true)
        .setSilent(true)
        .setOngoing(true)
        .setColor(getColor())
        .setSmallIcon(R.drawable.ic_logo_notification)
        .setPriority(NotificationCompat.PRIORITY_HIGH)
        .setVisibility(NotificationCompat.VISIBILITY_PUBLIC)
        .setForegroundServiceBehavior(NotificationCompat.FOREGROUND_SERVICE_IMMEDIATE)
        .build();
  }

  private int getColor() {
    Context colorContext;
    if (DynamicColors.isDynamicColorAvailable()) {
      colorContext = DynamicColors.wrapContextIfAvailable(context);
    } else {
      int themeResId;
      switch (sharedPrefs.getString(PREF.THEME, DEF.THEME)) {
        case THEME.RED:
          themeResId = R.style.Theme_Tack_Red;
          break;
        case THEME.GREEN:
          themeResId = R.style.Theme_Tack_Green;
          break;
        case THEME.BLUE:
          themeResId = R.style.Theme_Tack_Blue;
          break;
        default:
          themeResId = R.style.Theme_Tack_Yellow;
          break;
      }
      colorContext = new ContextThemeWrapper(context, themeResId);
    }
    return ResUtil.getSysColor(colorContext, R.attr.colorPrimary);
  }
}

File: app/src/main/kotlin/xyz/zedler/patrick/tack/util/NotificationUtil.kt
package xyz.zedler.patrick.tack.util

import android.Manifest
import android.annotation.SuppressLint
import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Build
import androidx.core.app.NotificationCompat
import androidx.core.app.NotificationCompat.Action
import androidx.core.app.NotificationManagerCompat
import androidx.core.content.ContextCompat
import xyz.zedler.patrick.tack.R
import xyz.zedler.patrick.tack.activity.MainActivity
import xyz.zedler.patrick.tack.service.MetronomeService
import kotlin.math.roundToInt

object NotificationUtil {

  const val NOTIFICATION_ID: Int = 1
  private const val CHANNEL_ID = "metronome"
  private const val REQUEST_CODE = 0

  fun hasPermission(context: Context): Boolean {
    return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
      ContextCompat.checkSelfPermission(
        context, Manifest.permission.POST_NOTIFICATIONS
      ) == PackageManager.PERMISSION_GRANTED
    } else {
      true
    }
  }

  fun createNotificationChannel(context: Context) {
    if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O) return
    val manager = context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
    manager.createNotificationChannel(
      NotificationChannel(
        CHANNEL_ID,
        context.getString(R.string.title_notification_channel),
        NotificationManager.IMPORTANCE_HIGH
      )
    )
  }

  @SuppressLint("MissingPermission")
  fun updateNotification(context: Context, notification: Notification) {
    if (hasPermission(context)) {
      NotificationManagerCompat.from(context).notify(NOTIFICATION_ID, notification)
    }
  }

  fun getNotification(
    context: Context,
    showPlayButton: Boolean,
    showTimer: Boolean,
    showPromotedLiveUpdate: Boolean,
    timerTextLong: String?,
    timerTextShort: String?,
    timerProgress: Float,
    timerDuration: Int
  ): Notification {
    val openIntent = Intent(context, MainActivity::class.java).apply {
      flags = Intent.FLAG_ACTIVITY_SINGLE_TOP or Intent.FLAG_ACTIVITY_REORDER_TO_FRONT
    }
    val activityPendingIntent = PendingIntent.getActivity(
      context, REQUEST_CODE, openIntent, PendingIntent.FLAG_IMMUTABLE
    )

    val startIntent = Intent(context, MetronomeService::class.java).apply {
      action = MetronomeService.ACTION_START
    }
    val startServicePendingIntent = PendingIntent.getService(
      context, REQUEST_CODE, startIntent, PendingIntent.FLAG_IMMUTABLE
    )
    val actionStart = Action(
      R.drawable.ic_rounded_play_arrow_fill,
      context.getString(R.string.action_play),
      startServicePendingIntent
    )

    val stopIntent = Intent(context, MetronomeService::class.java).apply {
      action = MetronomeService.ACTION_STOP
    }
    val stopServicePendingIntent = PendingIntent.getService(
      context, REQUEST_CODE, stopIntent, PendingIntent.FLAG_IMMUTABLE
    )
    val actionStop = Action(
      R.drawable.ic_rounded_stop_fill,
      context.getString(R.string.action_stop),
      stopServicePendingIntent
    )

    val dismissIntent = Intent(context, MetronomeService::class.java).apply {
      action = MetronomeService.ACTION_DISMISS
    }
    val dismissPendingIntent = PendingIntent.getService(
      context, REQUEST_CODE, dismissIntent, PendingIntent.FLAG_IMMUTABLE
    )

    val title = context.getString(R.string.msg_service_running)
    val text = context.getString(R.string.msg_service_running_return)

    val builder = NotificationCompat.Builder(context, CHANNEL_ID)
      .setContentTitle(title)
      .setContentText(text)
      .setStyle(NotificationCompat.BigTextStyle().setBigContentTitle(title).bigText(text))
      .setContentIntent(activityPendingIntent)
      .addAction(if (showPlayButton) actionStart else actionStop)
      .setAutoCancel(true)
      .setOnlyAlertOnce(true)
      .setSilent(true)
      .setOngoing(true)
      .setShowWhen(false)
      .setDeleteIntent(dismissPendingIntent)
      .setColor(getColor(context))
      .setSmallIcon(R.drawable.ic_logo_notification)
      .setPriority(NotificationCompat.PRIORITY_HIGH)
      .setVisibility(NotificationCompat.VISIBILITY_PUBLIC)
      .setForegroundServiceBehavior(NotificationCompat.FOREGROUND_SERVICE_IMMEDIATE)

    if (showTimer) {
      val progressStyle = NotificationCompat.ProgressStyle()
      val safeDuration = if (timerDuration > 0) timerDuration else 1

      if (safeDuration > 10) {
        progressStyle.setProgressSegments(
          listOf(NotificationCompat.ProgressStyle.Segment(100))
        )
      } else {
        val points = mutableListOf<NotificationCompat.ProgressStyle.Point>()
        val segments = mutableListOf<NotificationCompat.ProgressStyle.Segment>()
        var previousOffset = 0

        for (i in 1..safeDuration) {
          val currentOffset = ((i * 100f) / safeDuration).roundToInt()
          val segmentLength = currentOffset - previousOffset
          segments.add(NotificationCompat.ProgressStyle.Segment(segmentLength))
          if (safeDuration <= 5) {
            points.add(NotificationCompat.ProgressStyle.Point(currentOffset))
          }
          previousOffset = currentOffset
        }
        progressStyle.setProgressSegments(segments)
        if (points.isNotEmpty()) {
          progressStyle.setProgressPoints(points)
        }
      }
      progressStyle.setProgress((timerProgress * 100).toInt())

      builder.setContentText(timerTextLong)
        .setShortCriticalText(timerTextShort)
        .setStyle(progressStyle)
        .setRequestPromotedOngoing(showPromotedLiveUpdate)
    }
    return builder.build()
  }

  private fun getColor(context: Context): Int {
    return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
      ContextCompat.getColor(context, android.R.color.system_accent1_600)
    } else {
      ContextCompat.getColor(context, R.color.yellow_theme_primary)
    }
  }
}
